sap_business_document_processing/document_information_extraction_client/dox_api_client.py

```diff
@@ -0,0 +1,608 @@
+        if not silent:
+            upload_results = self._validate_results(upload_results, 'Some documents could not be uploaded successfully')
+
+        self.logger.info(f'Finished uploading {number_of_documents} documents for client {client_id}')
+        if not wait_for_results:
+            return self._create_result_iterator(upload_results)
+
+        return self.get_documents_results(upload_results, return_null_values, silent)
+
+    def get_documents_results(self, upload_results, return_null_values=False, silent=False):
+        """
+        Get results of multiple documents, given their upload results.
+        :param upload_results: A list of uploaded documents. This input parameter typically can be the output of method
+        upload_documents(..., wait_for_results=False).
+        :param return_null_values: Flag if fields with null as value should be included in the response or not.
+        Default is False
+        :param silent: If True, the functions returns even if some documents failed uploading or processing. If False,
+        will raise an exception if at least one document failed. Default is False
+        :return: An iterator with processed document results for successful documents and exceptions for failed
+        documents. Use next(iterator) within a try-catch block to filter the failed documents.
+        """
+        if not isinstance(upload_results, list):
+            upload_results = list(upload_results)
+        self.logger.debug(f'Waiting for {len(upload_results)} documents to be processed')
+        with ThreadPoolExecutor(min(self.polling_threads, len(upload_results))) as pool:
+            results = pool.map(self._get_document_wrap_errors, upload_results,
+                               [return_null_values] * len(upload_results))
+        if not silent:
+            results = self._validate_results(results, 'Some documents could not be processed successfully')
+        self.logger.info(f'{len(upload_results)} documents have been processed')
+        return self._create_result_iterator(results)
+
+    def _single_upload(self, document_path: str, options):
+        with open(document_path, 'rb') as file:
+            response = self.post(DOCUMENT_ENDPOINT,
+                                 files={API_REQUEST_FIELD_FILE: file},
+                                 data={API_REQUEST_FIELD_OPTIONS: json.dumps(options)})
+        return response.json()
+
+    def _single_upload_wrap_errors(self, document_path, options):
+        return function_wrap_errors(self._single_upload, document_path, options)
+
+    def _get_document_wrap_errors(self, upload_result, *args):
+        if isinstance(upload_result, Exception):
+            return upload_result
+        return function_wrap_errors(self.get_document_result, upload_result[API_FIELD_ID], True, *args)
+
+    def get_documents(self, client_id: str = None):
```

Comment:
Why do we prefer having `upload_documents`, `get_documents_results` and `get_documents` rather than `extract_information_from_document(s)`?

Reply:
I have now rearranged the functions to upload and process documents as follows:

- `extract_information_from_document`: uploads a single document and waits until the extraction result is done. The additional arguments are given as individual function parameters
- `extract_information_from_document_with_options`: as above, but the arguments are given as one `options` parameter
- `extract_information_from_documents`: uploads multiple documents and waits until all documents have been processed (or timeout is reached). Additional arguments are given as individual parameters
- `extract_information_from_documents_with_options`: as above, but the arguments are given as one `options` parameter
- `get_extraction_for_document`: retrieves the extracted information for a document given it's ID
- `get_extraction_for_documents`: retrieves the extracted information for multiple documents given their IDs
- `get_document_list`: gets a list of all documents that have been uploaded and their ID and current status. Does NOT return the extracted information. (can be filtered based on client ID)

Reply:
Looks very good, thanks!